apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/RdaLoadJob.java

```diff
@@ -0,0 +1,178 @@
+ *
+ * <p>Since the streaming service can run for extended periods of time this class is designed to be
+ * reentrant. If multiple threads invoke the call() method at the same time only the first thread
+ * will do any work. The other threads will all immediately return with an indication that they have
+ * no work to do.
+ */
+public final class RdaLoadJob<TResponse> implements PipelineJob<NullPipelineJobArguments> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RdaLoadJob.class);
+  public static final String CALLS_METER_NAME =
+      MetricRegistry.name(RdaLoadJob.class.getSimpleName(), "calls");
+  public static final String FAILURES_METER_NAME =
+      MetricRegistry.name(RdaLoadJob.class.getSimpleName(), "failures");
+  public static final String SUCCESSES_METER_NAME =
+      MetricRegistry.name(RdaLoadJob.class.getSimpleName(), "successes");
+  public static final String PROCESSED_METER_NAME =
+      MetricRegistry.name(RdaLoadJob.class.getSimpleName(), "processed");
+
+  private final Config config;
+  private final Callable<RdaSource<TResponse>> sourceFactory;
+  private final Callable<RdaSink<TResponse>> sinkFactory;
+  private final Meter callsMeter;
+  private final Meter failuresMeter;
+  private final Meter successesMeter;
+  private final Meter processedMeter;
+  // This is used to enforce that this job can only be executed by a single thread at any given
+  // time. If multiple threads call the job at the same time only the first will do any work.
+  private final Semaphore runningSemaphore;
+
+  public RdaLoadJob(
+      Config config,
+      Callable<RdaSource<TResponse>> sourceFactory,
+      Callable<RdaSink<TResponse>> sinkFactory,
+      MetricRegistry appMetrics) {
+    this.config = Preconditions.checkNotNull(config);
+    this.sourceFactory = Preconditions.checkNotNull(sourceFactory);
+    this.sinkFactory = Preconditions.checkNotNull(sinkFactory);
+    callsMeter = appMetrics.meter(CALLS_METER_NAME);
+    failuresMeter = appMetrics.meter(FAILURES_METER_NAME);
+    successesMeter = appMetrics.meter(SUCCESSES_METER_NAME);
+    processedMeter = appMetrics.meter(PROCESSED_METER_NAME);
+    runningSemaphore = new Semaphore(1);
+  }
+
+  @Override
+  public PipelineJobOutcome call() throws Exception {
+    // We only allow one outstanding call at a time.  If this job is already running any other
+    // call to the same job exits immediately with NOTHING_TO_DO.
+    if (!runningSemaphore.tryAcquire()) {
+      LOGGER.warn("job is already running");
+      return NOTHING_TO_DO;
+    }
+    try {
+      final long startMillis = System.currentTimeMillis();
+      int processedCount = 0;
+      Exception error = null;
+      try {
+        callsMeter.mark();
+        try (RdaSource<TResponse> source = sourceFactory.call();
+            RdaSink<TResponse> sink = sinkFactory.call()) {
+          processedCount = source.retrieveAndProcessObjects(config.getBatchSize(), sink);
```

Comment:
Yeah that's very possible.  If we need multiple sinks it will be easy to add a `CompoundRdaSink` that just invokes a series of sinks whenever it is invoked (from a `List` passed to it's constructor for example).  That would work nicely if the audit log is stored independently of the database.  If we wind up storing it in the database we might just let the `RdaSink` that writes the record also update the audit log directly using some class that knows about such things.